### ISSUE TITLE
Fast per example gradient support

### ISSUE THREAD

**Issue by AndreiBarsan** [tensorflow/privacy#86]
Section 4 of  the DP-SGD paper [0] described the method's implementation as leveraging the `per_example_gradient` operator described by Ian Goodfellow in [1], leading to overhead that just amounts to a "modest slowdown in training".

I ran the vectorized DP optimizer example on MNIST, but it still seems to take roughly 35s / epoch, as opposed to ~3s / epoch without DP-SGD. I am using a GTX 1080 GPU with CUDA 10.1 and driver version 430.50 on Ubuntu 18.04. This is the example I ran on my computer:
https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/optimizers/dp_optimizer_vectorized.py

This seems like a ~10x slowdown. Is this method actually leveraging the method from [1]? It's still substantially faster than a naive implementation I wrote in PyTorch, but the overhead still seems a bit high. Am I using the wrong example script for this, or does the nature of RMAD and its implementation in TF make it impossible to speed things up further?

I am running TensorFlow 2.0.0 and Python 3.7.

Thank you!

[0]: Abadi, M., McMahan, H. B., Chu, A., Mironov, I., Zhang, L., Goodfellow, I., & Talwar, K. (2016). Deep learning with differential privacy. Proceedings of the ACM Conference on Computer and Communications Security, 24-28-Octo(Ccs), 308–318. https://doi.org/10.1145/2976749.2978318
[1]: I. Goodfellow. Efficient per-example gradient computations. CoRR, abs/1510.01799v2, 2015.

**Comment by npapernot** [comment]
Thanks for the question. We moved away from the per example gradient optimization from [1] to be able to support a wider range of architectures and implementations. If you have any suggestions to provide further speed-ups, feel free to send me an email!

**Comment by AndreiBarsan** [comment]
Thank you for the clarification!
